tsl/src/remote/dist_copy.c

```diff
@@ -797,36 +1286,173 @@ send_copy_data(StringInfo row_data, const List *connections)
 	return true;
 }
 
+/*
+ * Rows for sending to a particular data node.
+ */
+typedef struct DataNodeRows
+{
+	int data_node_id;
+	Oid server_oid;
+	TSConnection *connection;
+	int rows_total;
+
+	/* Array of indices into the batch row array. */
+	int *row_indices;
+} DataNodeRows;
+
 static bool
 remote_copy_process_and_send_data(RemoteCopyContext *context)
 {
 	Hypertable *ht = context->ht;
-	Chunk *chunk;
-	Point *point;
-	const List *connections;
+	const int n = context->batch_row_count;
+	Assert(n <= MAX_BATCH_ROWS);
 
-	if (context->binary_operation)
-		point = get_current_point_for_binary_copy(context->data_context, ht->space);
-	else
-		point = get_current_point_for_text_copy(context->data_context);
+	/*
+	 * This list tracks the per-batch insert states of the data nodes
+	 * (DataNodeRows).
+	 */
+	List *data_nodes = NIL;
+
+	/* For each row, find or create the destination chunk. */
+	bool did_end_copy = false;
+	for (int row_in_batch = 0; row_in_batch < n; row_in_batch++)
+	{
+		Point *point = context->batch_points[row_in_batch];
+
+		Chunk *chunk = ts_hypertable_find_chunk_for_point(ht, point);
+		if (chunk == NULL)
+		{
+			if (!did_end_copy)
+			{
+				/*
+				 * The data node connections have to be flushed and the COPY
+				 * query ended before creating
+				 * a new chunk. They might have outstanding COPY data from the
+				 * previous batch.
+				 */
+				end_copy_on_success(&context->connection_state);
+				did_end_copy = true;
+			}
+			chunk = ts_hypertable_create_chunk_for_point(ht, point);
+		}
+
+		/*
+		 * For remote copy, we don't use chunk insert states on the AN.
+		 * So we need to explicitly set the chunk as unordered when copies
+		 * are directed to previously compressed chunks.
+		 */
+		if (ts_chunk_is_compressed(chunk) && (!ts_chunk_is_unordered(chunk)))
+			ts_chunk_set_unordered(chunk);
+
+		/*
+		 * Schedule the row for sending to the data nodes containing the chunk.
+		 */
+		ListCell *lc;
+		foreach (lc, chunk->data_nodes)
+		{
+			ChunkDataNode *chunk_data_node = lfirst(lc);
+			/* Find the existing insert state for this data node. */
+			DataNodeRows *data_node_rows = NULL;
+			ListCell *lc2;
+			foreach (lc2, data_nodes)
+			{
+				data_node_rows = lfirst(lc2);
+				if (chunk_data_node->foreign_server_oid == data_node_rows->server_oid)
+				{
+					break;
+				}
+			}
+
+			if (lc2 == NULL)
+			{
+				/* No insert state for this data node yet. Create it. */
+				data_node_rows = palloc(sizeof(DataNodeRows));
+				data_node_rows->connection = NULL;
+				data_node_rows->server_oid = chunk_data_node->foreign_server_oid;
+				data_node_rows->rows_total = 0;
+				data_node_rows->row_indices = palloc(sizeof(int) * context->batch_row_count);
```

Comment:
I assume that this is oversizing this array since not all tuples in the batch are sent to all nodes? Just checking, to ensure that we need not re-size this later or are undersizing, for example.

Reply:
Yep, added a comment.